use context managers for database cursor objects & stream results

arvidn opened the pull request:
To make sure this doesn't cause any performance regressions, I ran both the block_store and coin_store benchmarks.
I would have expected this patch to not be any worse, but maybe a little bit faster. using `async for` to stream the sqlite results back allows pipelining reading from disk and parsing.

Are these timings noise? Is the code change worth it anyway?

| test | before | after | after / before
---|---|---|---
| block_store v1 | 333.8625s | 339.1758s | 101.6 %
| block_store v2 | 306.5919s | 319.9539s | 104.4 %
| coin_store v1 | 228.9748s | 219.8814s | 96.0 %
| coin_store v2 | 193.9898s | 195.0622s | 100.6 %

## This patch

block_store
```
version 1
89.2887s, add_full_block
52.8452s, get_full_block
5.8553s, get_full_block_bytes
53.1792s, get_full_blocks_at
7.2331s, get_block_records_by_hash
54.2095s, get_blocks_by_hash
5.4857s, get_block_record
0.6965s, get_block_records_in_range
0.0057s, get_block_records_close_to_peak
4.2023s, get_block_record
66.1747s, get_random_not_compactified
all tests completed in 339.1758s
database size: 2418.156 MB
version 2
44.2622s, add_full_block
52.6760s, get_full_block
5.0684s, get_full_block_bytes
53.2927s, get_full_blocks_at
7.9133s, get_block_records_by_hash
53.9895s, get_blocks_by_hash
6.5078s, get_block_record
0.7906s, get_block_records_in_range
0.0066s, get_block_records_close_to_peak
4.4581s, get_block_record
90.9888s, get_random_not_compactified
all tests completed in 319.9539s
database size: 357.474 MB
```

coin_store
```
version 1
Building database 
49.3254s, MOSTLY ADDITIONS additions: 400400.0 removals: 20000.0
8.0602s, MOSTLY REMOVALS additions: 600 removals: 140000
67.6810s, FULLBLOCKS additions: 400400 removals: 400000
1.4604s, GET RECORDS BY NAMES with spent 200 lookups found 40000 coins in total
66.7356s, GET RECORDS BY NAMES without spent 200 lookups found 20641 coins in total
26.6187s, GET COINS REMOVED AT HEIGHT 800 blocks, found 580000 coins in total
all tests completed in 219.8814s
database size: 638.099 MB
version 2
Building database 
44.3007s, MOSTLY ADDITIONS additions: 400400.0 removals: 20000.0
7.6507s, MOSTLY REMOVALS additions: 600 removals: 140000
64.7445s, FULLBLOCKS additions: 400400 removals: 400000
2.3373s, GET RECORDS BY NAMES with spent 200 lookups found 40000 coins in total
51.2288s, GET RECORDS BY NAMES without spent 200 lookups found 20693 coins in total
24.8002s, GET COINS REMOVED AT HEIGHT 800 blocks, found 580000 coins in total
all tests completed in 195.0622s
database size: 353.669 MB
```

## Prior commit:

block_store:
```
version 1
91.3370s, add_full_block
52.0561s, get_full_block
5.4880s, get_full_block_bytes
50.7568s, get_full_blocks_at
5.7596s, get_block_records_by_hash
53.2185s, get_blocks_by_hash
7.0929s, get_block_record
0.6808s, get_block_records_in_range
0.0057s, get_block_records_close_to_peak
4.8409s, get_block_record
62.6261s, get_random_not_compactified
all tests completed in 333.8625s
database size: 2418.156 MB
version 2
45.1364s, add_full_block
51.2920s, get_full_block
5.7642s, get_full_block_bytes
49.4884s, get_full_blocks_at
6.8339s, get_block_records_by_hash
52.1692s, get_blocks_by_hash
6.5498s, get_block_record
0.6922s, get_block_records_in_range
0.0062s, get_block_records_close_to_peak
4.2534s, get_block_record
84.4062s, get_random_not_compactified
all tests completed in 306.5919s
database size: 357.474 MB
```

coin_store
```
version 1
Building database 
56.8906s, MOSTLY ADDITIONS additions: 400400.0 removals: 20000.0
8.2210s, MOSTLY REMOVALS additions: 600 removals: 140000
70.9838s, FULLBLOCKS additions: 400400 removals: 400000
1.3132s, GET RECORDS BY NAMES with spent 200 lookups found 40000 coins in total
67.1411s, GET RECORDS BY NAMES without spent 200 lookups found 20641 coins in total
24.4251s, GET COINS REMOVED AT HEIGHT 800 blocks, found 580000 coins in total
all tests completed in 228.9748s
database size: 638.091 MB
version 2
Building database 
44.5126s, MOSTLY ADDITIONS additions: 400400.0 removals: 20000.0
7.9520s, MOSTLY REMOVALS additions: 600 removals: 140000
63.4308s, FULLBLOCKS additions: 400400 removals: 400000
2.1089s, GET RECORDS BY NAMES with spent 200 lookups found 40000 coins in total
49.9227s, GET RECORDS BY NAMES without spent 200 lookups found 20693 coins in total
26.0627s, GET COINS REMOVED AT HEIGHT 800 blocks, found 580000 coins in total
all tests completed in 193.9898s
database size: 353.620 MB
```

arvidn committed:
use context managers for database cursor objects, to ensure they are closed. Also stream results back from cursors, rather than reading the full list of results up-front

arvidn commented:
I'm having second thoughts about using `async for` to stream results back. I suspect that might contributing to slowing this down. It might add more context switches for instance.

altendky reviewed (changes requested):
Some tweaks, but otherwise good.  If you remove the async iteration, that makes it a bit more focused of a PR so...  either way.

I did do a bit of benchmarking myself and found about a 1% apparent slow down though there was 2-3% stddev on the actual runtime of either leaving me a bit skeptical of the performance importance.  Though, I didn't dig out how to conclude if there was a statistically significant difference or not.

https://gist.github.com/altendky/6b7b564f60ed8a4cc1c7e24945f6b45a

https://docs.google.com/spreadsheets/d/1ygeCnuYJA7IfI5A_BMi-6H99_sFAyUetlWAUo-Rnz1M/edit#gid=0



| before | after |  
| -- | -- | -- |
| 274.3248 | 292.738 | 6.7%
| 271.7372 | 274.6443 | 1.1%
| 288.1356 | 291.053 | 1.0%
| 287.8522 | 291.7024 | 1.3%
| 292.1706 | 291.653 | -0.2%
| 291.8701 | 292.7326 | 0.3%
| 288.388 | 290.962 | 0.9%
| 288.0278 | 293.2307 | 1.8%
|  |   |  
| 285.3132875 | 289.8395 |  
| 7.803895491 | 6.195129856 |  
| 2.7% | 2.1% |  

arvidn commented on chia/full_node/block_height_map.py:
```diff
@@ -58,30 +58,25 @@ async def create(cls, blockchain_dir: Path, db: DBWrapper) -> "BlockHeightMap":
-                return self
+            async with self.db.db.execute("SELECT hash FROM current_peak WHERE key = 0") as cursor:
+                peak_row = await cursor.fetchone()
+                if peak_row is None:
+                    return self
```
yeah. the reason I did that was that it still feels strange to me that variables declared in an inner scope still exist outside of the scope..

altendky replied:
It's not a scope.  `:]`  You get scopes for a module, class (though these are... weird), function, comprehensions...  and lambdas i guess.

arvidn commented on chia/full_node/block_store.py:
```diff
@@ -151,10 +151,9 @@ async def add_full_block(self, header_hash: bytes32, block: FullBlock, block_rec
                     bytes(block_record),
                 ),
             )
-            await cursor_1.close()
```
I modeled this after the first example from the docs: https://aiosqlite.omnilib.dev/en/stable/index.html

```
await db.execute("INSERT INTO some_table ...")
```

arvidn commented on chia/full_node/block_store.py:
```diff
@@ -263,10 +257,11 @@ async def get_full_blocks_at(self, heights: List[uint32]) -> List[FullBlock]:
-        await cursor.close()
-        return [self.maybe_decompress(row[0]) for row in rows]
+        async with self.db.execute(formatted_str, heights_db) as cursor:
+            ret: List[FullBlock] = []
+            async for row in cursor:
+                ret.append(self.maybe_decompress(row[0]))
+            return ret
```
I think I'll back-out `async for` from this patch for now though

arvidn commented:
@altendky I dropped all the `async for` and incorporated your suggested changes. The new timings are closer (but I just ran the new ones once).

This commit:

block_store:
```
version 1
89.6409s, add_full_block
53.2567s, get_full_block
5.2159s, get_full_block_bytes
51.1661s, get_full_blocks_at
5.8824s, get_block_records_by_hash
52.6979s, get_blocks_by_hash
7.1451s, get_block_record
0.6540s, get_block_records_in_range
0.0053s, get_block_records_close_to_peak
4.6177s, get_block_record
66.4469s, get_random_not_compactified
all tests completed in 336.7288s
database size: 2418.156 MB
version 2
44.0412s, add_full_block
52.3799s, get_full_block
5.0736s, get_full_block_bytes
50.2342s, get_full_blocks_at
6.3605s, get_block_records_by_hash
53.0350s, get_blocks_by_hash
6.6490s, get_block_record
0.7128s, get_block_records_in_range
0.0061s, get_block_records_close_to_peak
4.5579s, get_block_record
88.9816s, get_random_not_compactified
all tests completed in 312.0317s
database size: 357.474 MB
```

coin_store:
```
version 1
Building database
52.9241s, MOSTLY ADDITIONS additions: 400400.0 removals: 20000.0
7.6849s, MOSTLY REMOVALS additions: 600 removals: 140000
72.3118s, FULLBLOCKS additions: 400400 removals: 400000
1.4111s, GET RECORDS BY NAMES with spent 200 lookups found 40000 coins in total
68.3960s, GET RECORDS BY NAMES without spent 200 lookups found 20641 coins in total
24.7959s, GET COINS REMOVED AT HEIGHT 800 blocks, found 580000 coins in total
all tests completed in 227.5239s
database size: 638.095 MB
version 2
Building database
45.5893s, MOSTLY ADDITIONS additions: 400400.0 removals: 20000.0
7.9858s, MOSTLY REMOVALS additions: 600 removals: 140000
64.8043s, FULLBLOCKS additions: 400400 removals: 400000
2.1198s, GET RECORDS BY NAMES with spent 200 lookups found 40000 coins in total
50.4896s, GET RECORDS BY NAMES without spent 200 lookups found 20693 coins in total
23.2696s, GET COINS REMOVED AT HEIGHT 800 blocks, found 580000 coins in total
all tests completed in 194.2584s
database size: 353.620 MB
```

altendky commented on chia/full_node/block_store.py:
```diff
@@ -481,48 +458,43 @@ async def set_peak(self, header_hash: bytes32) -> None:
-        heights = []
-        for row in rows:
-            heights.append(int(row[0]))
+            heights = []
+            for row in rows:
+                heights.append(int(row[0]))
```
https://github.com/Chia-Network/chia-blockchain/runs/4571820373?check_suite_focus=true

```suggestion
        heights = [int(row[0]) for row in rows]
```

arvidn committed:
Update chia/full_node/block_store.py

Co-authored-by: Kyle Altendorf <sda@fstab.net>

altendky reviewed (approved):
I would generally like to see the minimal amount of code in the context blocks as possible so resources get released quickly.  As always, happy to chat about it.  But, I don't think the cases here are significant issues so not getting hung up on that now.

There were some timeouts on the tests for the plotting directory on Ubuntu but the timeout rate on this PR and main seem the same as of running them both now.  I guess I'll have to dig into that separately.  I haven't seen the hangs locally yet.

Thanks for the context management!

arvidn commented:
Although that's probably a good rule of thumb, when it comes to async processes, I don't think it's necessarily always true. For instance, perhaps the exiting of the context will synchronize (and wait for) some background task completing. If you can defer this synchronization, you may increase concurrency.

But perhaps there aren't any examples of that in the real world.
